resolver.go

```diff
@@ -490,35 +490,51 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 				continue
 			}
 			r.forwardQueryEnd()
-			if resp != nil {
-				if resp.Rcode == dns.RcodeServerFailure {
-					// for Server Failure response, continue to the next external DNS server
-					logrus.Debugf("[resolver] external DNS %s:%s responded with ServFail for %q", proto, extDNS.IPStr, name)
-					continue
-				}
-				answers := 0
-				for _, rr := range resp.Answer {
-					h := rr.Header()
-					switch h.Rrtype {
-					case dns.TypeA:
-						answers++
-						ip := rr.(*dns.A).A
-						logrus.Debugf("[resolver] received A record %q for %q from %s:%s", ip, h.Name, proto, extDNS.IPStr)
-						r.backend.HandleQueryResp(h.Name, ip)
-					case dns.TypeAAAA:
-						answers++
-						ip := rr.(*dns.AAAA).AAAA
-						logrus.Debugf("[resolver] received AAAA record %q for %q from %s:%s", ip, h.Name, proto, extDNS.IPStr)
-						r.backend.HandleQueryResp(h.Name, ip)
-					}
-				}
-				if resp.Answer == nil || answers == 0 {
-					logrus.Debugf("[resolver] external DNS %s:%s did not return any %s records for %q", proto, extDNS.IPStr, queryType, name)
-				}
-				resp.Compress = true
-			} else {
+
+			if resp == nil {
 				logrus.Debugf("[resolver] external DNS %s:%s returned empty response for %q", proto, extDNS.IPStr, name)
+				break
+			}
+			switch resp.Rcode {
+			case dns.RcodeServerFailure:
+				// Server Failure response: continue to the next external DNS server
+				logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), name)
+				continue
+			case dns.RcodeRefused:
+				// Server returned REFUSED: this can be a transitional status, so continue with the next external DNS server
+				logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), name)
+				continue
+			case dns.RcodeNameError:
+				// Server returned NXDOMAIN. Stop resolution: it's an authoritative answer (see RFC 8020: https://tools.ietf.org/html/rfc8020#section-2)
+				logrus.Debugf("[resolver] external DNS %s:%s responded with %s for %q", proto, extDNS.IPStr, statusString(resp.Rcode), name)
+				break
+			case dns.RcodeSuccess:
+				// All is well
+			default:
+				// TODO should we log as error, and break here?
+				// Server gave some error. Log the error, but proceed as usual
+				logrus.Debugf("[resolver] external DNS %s:%s responded with %s (code %d) for %q", proto, extDNS.IPStr, statusString(resp.Rcode), resp.Rcode, name)
```

Comment:
we should probably handle all codes to match the RFC.

Reply:
I notice above that the branches for `RcodeServerFailure` and `RcodeRefused` are identical except for the comments.  I guess that's OK since they call out specific rationale for the decision, but they could be collapsed into the `default:` branch w/ no loss of information.  (just an extra code # in the log).

So, then took a look at https://tools.ietf.org/html/rfc6895 to see what codes were missing and there are a bunch.   For most of them, I think the safe thing to do would just be log and `continue` on to the next server's query.  So this would mean putting `continue` in the `default` branch above.  Should be safe, for example, to go on to the next query when we get `FormErr`, `NotImp` (should never get either of these), `Refused`, (another server might not refuse)  and `NotAuth` (another server might be authoritative).  The codes `YXDomain`, `YXRRSet`, `NXRRSet`, `NotZone` seem to indicate that the query is for a bad name.  If so, `continue`ing is safe in a sense but doesn't return good error info to the container.  We should never get any of the codes > 16 because we aren't doing DNSSEC, but if we did, we certainly shouldn't accept anything in the answer section from them.

TLDR:
  * the default branch should at least `continue` so we don't accept answers w/ potential error codes
  * we could collapse `RcodeRefused` and `RcodeServerFailure` into said branch but can see leaving it explicit
  * more robust handling would return proper responses for some of these errors but that could be left for another PR